waiter/src/waiter/scheduler.clj

```diff
@@ -642,3 +647,192 @@
           "WAITER_SERVICE_ID" service-id
           "WAITER_USERNAME" "waiter"}))
+
+;; Support for tracking service instance launching time stats
+
+(def instance-counts-zero {:requested 0 :running 0})
+
+(defn- make-launch-tracker
+  ([service-id service-id->service-description-fn]
+   (make-launch-tracker service-id service-id->service-description-fn instance-counts-zero #{}))
+  ([service-id service-id->service-description-fn instance-counts known-instance-ids]
+   {:instance-counts instance-counts
+    :instance-scheduling-start-times []
+    :known-instance-ids known-instance-ids
+    :metric-group (-> service-id service-id->service-description-fn (get "metric-group"))
+    :service-schedule-timer (metrics/service-timer service-id "launch-overhead" "schedule-time")
+    :service-startup-timer (metrics/service-timer service-id "launch-overhead" "startup-time")
+    :starting-instance-id->start-timestamp {}}))
+
+(defn update-launch-trackers
+  "Track metrics on launching service instances, specifically:
+   1) the time from requesting a new instance until it is scheduled (the \"schedule\" time), and
+   2) the time from scheduling a new instance until it becomes healthy (the \"startup\" time)."
+  [service-id->launch-tracker new-service-ids removed-service-ids service-id->healthy-instances
+   service-id->unhealthy-instances service-id->instance-counts
+   service-id->service-description-fn waiter-schedule-timer]
+  (let [current-time (t/now)
+        ;; Remove deleted services and add newly-discovered services
+        service-id->launch-tracker'
+        (merge (reduce dissoc service-id->launch-tracker removed-service-ids)
+               (pc/map-from-keys #(make-launch-tracker % service-id->service-description-fn)
+                                 new-service-ids))]
+    ;; Update all trackers...
+    (pc/for-map [[service-id
+                  {:keys [instance-counts instance-scheduling-start-times known-instance-ids
+                          metric-group service-schedule-timer service-startup-timer
+                          starting-instance-id->start-timestamp] :as tracker-state}]
+                 service-id->launch-tracker']
+      service-id
+      (let [healthy-instances (get service-id->healthy-instances service-id)
+            healthy-instance-id? (->> healthy-instances (map :id) set)
+            known-instances' (->> service-id
+                                  (get service-id->unhealthy-instances)
+                                  (concat healthy-instances))
+            previously-known-instance? (comp known-instance-ids :id)
+            new-instance-ids (->> known-instances'
+                                  (remove previously-known-instance?)
+                                  (sort instance-comparator)
+                                  (mapv :id))
+            known-instance-ids' (->> known-instances' (map :id) set)
+            removed-instance-ids (remove known-instance-ids known-instance-ids')
```

Comment:
Nitpick: `(set/difference known-instance-ids known-instance-ids')` is easier to read for me and returns a set.

Reply:
Done.